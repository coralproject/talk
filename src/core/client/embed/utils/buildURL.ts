export default function buildURL({
  protocol = window.location.protocol,
  hostname = window.location.hostname,
  port = window.location.port,
  pathname = window.location.pathname,
  search = window.location.search,
  hash = window.location.hash,
} = {}) {
  if (search && search[0] !== "?") {
    search = `?${search}`;
  } else if (search === "?") {
    search = "";
  }
  return `${protocol}//${hostname}${
    port ? `:${port}` : ""
  }${pathname}${search}${hash}`;
}
