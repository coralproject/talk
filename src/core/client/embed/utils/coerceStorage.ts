import { createInMemoryStorage } from "coral-framework/lib/storage";

function coerceStorage(type: "localStorage" | "sessionStorage"): Storage {
  let storage: Storage;
  try {
    storage = window[type];
  } catch {
    // eslint-disable-next-line no-console
    console.debug(`${type} not available, use in-memory replacement`);

    storage = createInMemoryStorage();
  }

  return storage;
}

export default coerceStorage;
