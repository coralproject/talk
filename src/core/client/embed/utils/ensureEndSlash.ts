export default function ensureEndSlash(p: string) {
  return p.match(/\/$/) ? p : `${p}/`;
}
