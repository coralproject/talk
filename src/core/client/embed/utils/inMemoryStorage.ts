class InMemoryStorage implements Storage {
  private readonly data = new Map<string, string>();

  public get length() {
    return this.data.size;
  }

  public clear() {
    this.data.clear();
  }

  public key(n: number) {
    const keys = [...this.data.keys()];
    if (keys.length < n || n < 0) {
      return null;
    }

    return keys[n];
  }

  public getItem(key: string): string | null {
    return this.data.get(key) || null;
  }

  public removeItem(key: string): void {
    this.data.delete(key);
  }

  public setItem(key: string, value: string): void {
    this.data.set(key, value);
  }
}

export default InMemoryStorage;
