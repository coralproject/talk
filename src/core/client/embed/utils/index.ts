export { default as buildURL } from "./buildURL";
export { default as ensureEndSlash } from "./ensureEndSlash";
export { default as ensureNoEndSlash } from "./ensureNoEndSlash";
export { default as startsWith } from "./startsWith";
export { default as prefixStorage } from "./prefixStorage";
export { default as parseHashQuery } from "./parseHashQuery";
