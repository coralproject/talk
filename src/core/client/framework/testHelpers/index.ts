export {
  default as createRelayEnvironment,
  CreateRelayEnvironmentParams,
} from "./createRelayEnvironment";
export { default as createFluentBundle } from "./createFluentBundle";
export { default as createSinonStub } from "./createSinonStub";
export {
  default as removeFragmentRefs,
  NoFragmentRefs,
} from "./removeFragmentRefs";
export { default as createUUIDGenerator } from "./createUUIDGenerator";
export * from "./denormalize";
export { default as replaceHistoryLocation } from "./replaceHistoryLocation";
