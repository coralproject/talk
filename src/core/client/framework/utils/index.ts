export { default as buildURL } from "./buildURL";
export { default as parseURL } from "./parseURL";
export { default as modifyQuery } from "./modifyQuery";
