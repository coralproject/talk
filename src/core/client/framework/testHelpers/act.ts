import TestRenderer from "react-test-renderer";

import { isPromiseLike } from "coral-common/utils";

export default function act<T>(callback: () => T): T {
  let callbackResult: T;
  const actResult = TestRenderer.act(() => {
    callbackResult = callback();
    return callbackResult as any;
  });
  if (isPromiseLike(callbackResult!)) {
    // Return it this way, to preserve warnings that React emits.
    return {
      then(resolve: (value: any) => void, reject: (err?: Error) => void) {
        (actResult as PromiseLike<any>).then(() => {
          resolve(callbackResult);
        }, reject);
      },
    } as any;
  }
  return callbackResult!;
}
