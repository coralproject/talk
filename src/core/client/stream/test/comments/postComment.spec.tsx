import { ReactTestRenderer } from "react-test-renderer";
import sinon from "sinon";
import timekeeper from "timekeeper";

import { timeout } from "talk-common/utils";
import { createSinonStub } from "talk-framework/testHelpers";

import { assets, baseComment, settings, users } from "../fixtures";
import create from "./create";

let testRenderer: ReactTestRenderer;
beforeEach(() => {
  const resolvers = {
    Query: {
      settings: sinon.stub().returns(settings),
      me: sinon.stub().returns(users[0]),
      asset: createSinonStub(
        s => s.throws(),
        s =>
          s
            .withArgs(undefined, { id: assets[0].id, url: null })
            .returns(assets[0])
      ),
    },
    Mutation: {
      createComment: createSinonStub(
        s => s.throws(),
        s =>
          s
            .withArgs(undefined, {
              input: {
                assetID: assets[0].id,
                body: "<strong>Hello world!</strong>",
                clientMutationId: "0",
              },
            })
            .returns({
              // TODO: add a type assertion here to ensure that if the type changes, that the test will fail
              edge: {
                cursor: null,
                node: {
                  ...baseComment,
                  id: "comment-x",
                  author: users[0],
                  body: "<strong>Hello world! (from server)</strong>",
                },
              },
              clientMutationId: "0",
            })
      ),
    },
  };

  ({ testRenderer } = create({
    // Set this to true, to see graphql responses.
    logNetwork: false,
    resolvers,
    initLocalState: localRecord => {
      localRecord.setValue(assets[0].id, "assetID");
    },
  }));
});

it("renders comment stream", async () => {
  // Wait for loading.
  await timeout();
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

it("post a comment", async () => {
  // Wait for loading.
  await timeout();
  testRenderer.root
    .findByProps({ inputId: "comments-postCommentForm-field" })
    .props.onChange({ html: "<strong>Hello world!</strong>" });

  timekeeper.freeze(new Date(baseComment.createdAt));

  testRenderer.root
    .findByProps({ id: "comments-postCommentForm-form" })
    .props.onSubmit();

  timekeeper.reset();

  // Test optimistic response.
  expect(testRenderer.toJSON()).toMatchSnapshot("optimistic response");

  // Wait for loading.
  await timeout();

  // Test after server response.
  expect(testRenderer.toJSON()).toMatchSnapshot("server response");
});
