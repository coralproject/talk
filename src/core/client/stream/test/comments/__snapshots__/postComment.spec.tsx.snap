// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post a comment: optimistic response 1`] = `
<div
  data-testid="comment-uuid-0"
>
  <div
    className="HorizontalGutter-root HorizontalGutter-full"
  >
    <div
      className="IndentedComment-blur Indent-root"
    >
      <div
        className=""
      >
        <div
          className="Comment-root"
          role="article"
        >
          <div
            className="Flex-root Flex-flex Flex-justifySpaceBetween Flex-directionRow"
          >
            <div
              className="Flex-root Flex-flex Flex-halfItemGutter Flex-alignBaseline Flex-directionColumn"
            >
              <span
                className="Typography-root Typography-heading3 Typography-colorTextPrimary Username-root"
              >
                Markus
              </span>
              <div
                className="Flex-root Flex-flex Flex-itemGutter Flex-alignBaseline Flex-directionRow"
              >
                <time
                  className="Timestamp-root RelativeTime-root"
                  dateTime="2018-07-06T18:24:00.000Z"
                  title="2018-07-06T18:24:00.000Z"
                >
                  2018-07-06T18:24:00.000Z
                </time>
              </div>
            </div>
            <div>
              <button
                className="BaseButton-root Button-root Button-sizeRegular Button-colorPrimary Button-variantUnderlined"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                type="button"
              >
                Edit
              </button>
            </div>
          </div>
          <div
            className="HorizontalGutter-root HorizontalGutter-full"
          >
            <div
              className="HTMLContent-root"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<b>Hello world!</b>",
                }
              }
            />
            <div
              className="Flex-root Flex-flex Flex-justifySpaceBetween"
            >
              <div
                className="Flex-root Flex-flex Flex-halfItemGutter Flex-directionRow"
              >
                <button
                  className="BaseButton-root Button-root Button-sizeSmall Button-colorRegular Button-variantGhost"
                  disabled={false}
                  id="comments-commentContainer-replyButton-uuid-0"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onMouseOut={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  type="button"
                >
                  <span>
                    Reply
                  </span>
                </button>
                <div
                  className="Popover-root"
                >
                  <button
                    aria-controls="permalink-popover-uuid-0"
                    className="BaseButton-root Button-root Button-sizeSmall Button-colorRegular Button-variantGhost"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    onTouchEnd={[Function]}
                    type="button"
                  >
                    <span>
                      Share
                    </span>
                  </button>
                  <div
                    aria-hidden={true}
                    aria-labelledby="permalink-popover-uuid-0-ariainfo"
                    id="permalink-popover-uuid-0"
                    role="popup"
                  >
                    <div
                      className="AriaInfo-root"
                      id="permalink-popover-uuid-0-ariainfo"
                    >
                      A dialog showing a permalink to the comment
                    </div>
                  </div>
                </div>
                <button
                  className="BaseButton-root Button-root Button-sizeSmall Button-colorRegular Button-variantGhost"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onMouseOut={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  type="button"
                >
                  <span>
                    Respect
                  </span>
                </button>
              </div>
              <div
                className="Flex-root Flex-flex Flex-halfItemGutter Flex-directionRow"
              >
                <div
                  className="Popover-root"
                >
                  <button
                    aria-controls="report-popover-uuid-0"
                    className="BaseButton-root Button-root Button-sizeSmall Button-colorRegular Button-variantGhost"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    onTouchEnd={[Function]}
                    type="button"
                  >
                    <span>
                      Report
                    </span>
                  </button>
                  <div
                    aria-hidden={true}
                    aria-labelledby="report-popover-uuid-0-ariainfo"
                    id="report-popover-uuid-0"
                    role="popup"
                  >
                    <div
                      className="AriaInfo-root"
                      id="report-popover-uuid-0-ariainfo"
                    >
                      A dialog for reporting comments
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
